encodings/fastlanes/src/bitpacking/compress.rs

```diff
@@ -196,39 +200,52 @@ pub fn gather_patches(
 }
 
 pub fn unpack(array: BitPackedArray) -> VortexResult<PrimitiveArray> {
+    match_each_integer_ptype_with_unsigned_type!(array.ptype(), |$P, $UnsignedT| {
+        unpack_primitive::<$P, $UnsignedT>(array)
+    })
+}
+
+pub fn unpack_primitive<T: NativePType, UnsignedT: NativePType + BitPacking>(
+    array: BitPackedArray,
+) -> VortexResult<PrimitiveArray> {
+    let mut builder =
+        PrimitiveBuilder::with_capacity(array.dtype().nullability(), array.len() + 1024);
+    assert!(size_of::<T>() == size_of::<UnsignedT>());
+    unpack_into::<T, UnsignedT, _, _>(
+        array,
+        &mut builder,
+        |x| unsafe { std::mem::transmute(x) },
+        |x| unsafe { std::mem::transmute(x) },
+    )?;
+    builder.finish_into_primitive()
+}
+
+pub(crate) fn unpack_into<T: NativePType, UnsignedT: NativePType + BitPacking, F, G>(
+    array: BitPackedArray,
+    // TODO(ngates): do we want to use fastlanes alignment for this buffer?
+    builder: &mut PrimitiveBuilder<T>,
+    transmute: F,
+    transmute_mut: G,
+) -> VortexResult<()>
+where
+    F: Fn(&[UnsignedT]) -> &[T],
+    G: Fn(&mut [T]) -> &mut [UnsignedT],
+{
     let bit_width = array.bit_width() as usize;
     let length = array.len();
     let offset = array.offset() as usize;
-    let ptype = array.ptype();
-    let mut unpacked = match_each_unsigned_integer_ptype!(ptype.to_unsigned(), |$P| {
-        PrimitiveArray::new(
-            unpack_primitive::<$P>(array.packed_slice::<$P>(), bit_width, offset, length),
-            array.validity(),
-        )
-    });
-
-    // Cast to signed if necessary
-    if ptype.is_signed_int() {
-        unpacked = unpacked.reinterpret_cast(ptype);
-    }
 
-    if let Some(patches) = array.patches() {
-        unpacked.patch(patches)
-    } else {
-        Ok(unpacked)
-    }
-}
+    builder.nulls.append_validity(array.validity(), length)?;
 
-pub fn unpack_primitive<T: NativePType + BitPacking>(
-    packed: &[T],
-    bit_width: usize,
-    offset: usize,
-    length: usize,
-) -> Buffer<T> {
     if bit_width == 0 {
-        return buffer![T::zero(); length];
+        builder.append_zeros(length);
+        return Ok(());
     }
 
+    let packed = array.packed_slice::<UnsignedT>();
+    let builder_current_length = builder.len();
+    builder.values.reserve((length + 1023) / 1024 * 1024);
```

Comment:
So do you think this is the part that reallocated and effects perf?

Reply:
We know the length of the array, so could you instead allocated a 1024 * |T| and decode the last block into that and then copy values from this to the builder? (If there is not already enough space).

Reply:
In the benchmarks, we ensure the resulting builder is big enough to hold all the elements plus any scratch space needed by the decoded (in this case, 1024 empty spots, though only 1023 are needed). This operation shouldn't allocate in the benchmarks. I've verified this in the profile (there's no `BufferMut::reserve` or `BytesMut::extend_from_slice` both of which I've seen when it allocates).

Yeah, what you're describing is a reasonable alternative to adding extra space to the builders. I'd be happy to do that, though at this point it feels like we should do that in a follow up PR since this one is fairly large already.